common/src/main/scala/org/neo4j/spark/converter/DataConverter.scala

```diff
@@ -0,0 +1,181 @@
+package org.neo4j.spark.converter
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{GenericRowWithSchema, UnsafeArrayData, UnsafeMapData, UnsafeRow}
+import org.apache.spark.sql.catalyst.util.{ArrayBasedMapData, ArrayData, DateTimeUtils}
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+import org.neo4j.driver.internal._
+import org.neo4j.driver.{Value, Values}
+import org.neo4j.spark.service.SchemaService
+import org.neo4j.spark.util.Neo4jUtil
+
+import java.time._
+import java.time.format.DateTimeFormatter
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+trait DataConverter[T] {
+  def convert(value: Any, dataType: DataType = null): T
+
+  @tailrec
+  private[converter] final def extractStructType(dataType: DataType): StructType = dataType match {
+    case structType: StructType => structType
+    case mapType: MapType => extractStructType(mapType.valueType)
+    case arrayType: ArrayType => extractStructType(arrayType.elementType)
+    case _ => throw new UnsupportedOperationException(s"$dataType not supported")
+  }
+}
+
+object SparkToNeo4jDataConverter {
+  def apply(): SparkToNeo4jDataConverter = new SparkToNeo4jDataConverter()
+}
+
+class SparkToNeo4jDataConverter extends DataConverter[Value] {
+  override def convert(value: Any, dataType: DataType): Value = {
+    value match {
+      case date: java.sql.Date => convert(date.toLocalDate, dataType)
+      case timestamp: java.sql.Timestamp => convert(timestamp.toLocalDateTime, dataType)
+      case intValue: Int if dataType == DataTypes.DateType => convert(DateTimeUtils
+        .toJavaDate(intValue), dataType)
+      case longValue: Long if dataType == DataTypes.TimestampType => convert(DateTimeUtils
+        .toJavaTimestamp(longValue), dataType)
+      case unsafeRow: UnsafeRow => {
```

Comment:
maybe worth adding a comment with concrete examples where this `UnsafeRow` is useful